pkg/cloud/gcp/actuators/machine/reconciler.go

```diff
@@ -39,6 +38,91 @@ func newReconciler(scope *machineScope) *Reconciler {
 	}
 }
 
+var (
+	supportedGpuTypes = map[string]string{
+		"nvidia-tesla-k80":  "NVIDIA_K80_GPUS",
+		"nvidia-tesla-p100": "NVIDIA_P100_GPUS",
+		"nvidia-tesla-v100": "NVIDIA_V100_GPUS",
+		"nvidia-tesla-a100": "NVIDIA_A100_GPUS",
+		"nvidia-tesla-p4":   "NVIDIA_P4_GPUS",
+		"nvidia-tesla-t4":   "NVIDIA_T4_GPUS",
+	}
+)
+
+func containsString(sli []string, str string) bool {
+	for _, elem := range sli {
+		if elem == str {
+			return true
+		}
+	}
+	return false
+}
+
+func (r *Reconciler) checkQuota(machineFamily int64) error {
+	region, err := r.computeService.RegionGet(r.projectID, r.providerSpec.Region)
+	if err != nil {
+		return fmt.Errorf("failed to get region via compute service: %v", err)
+	}
+	quotas := region.Quotas
+	var guestAccelerators = []*v1beta1.GCPAcceleratorConfig{}
+	if machineFamily != 0 {
+		guestAccelerators = append(guestAccelerators, &v1beta1.GCPAcceleratorConfig{AcceleratorType: "nvidia-tesla-a100", AcceleratorCount: machineFamily})
+	} else {
+		guestAccelerators = r.providerSpec.GuestAccelerators
+	}
+	// validate zone and then quota
+	for _, elem := range guestAccelerators {
+		_, err := r.computeService.AcceleratorTypesList(r.projectID, r.providerSpec.Zone, elem.AcceleratorType)
+		if err != nil {
+			return fmt.Errorf("AcceleratorType not available in the zone: %v", err)
+		}
+		metric := supportedGpuTypes[elem.AcceleratorType]
+		if metric == "" {
+			return machinecontroller.InvalidMachineConfiguration("Unsupported accelerator type")
+		}
+		// preemptible instances have separate quota
+		if r.providerSpec.Preemptible {
+			metric = "PREEMPTIBLE_" + metric
+		}
+		// check quota for GA
+		for i, q := range quotas {
+			if q.Metric == metric {
+				if int64(q.Usage)+elem.AcceleratorCount > int64(q.Limit) {
+					return machinecontroller.InvalidMachineConfiguration(fmt.Sprintf("Quota exceeded. Metric: %s. Usage: %v. Limit: %v.", metric, q.Usage, q.Limit))
+				}
+				break
+			}
+			if i == len(quotas)-1 {
+				return machinecontroller.InvalidMachineConfiguration(fmt.Sprintf("No quota found. Metric: %s.", metric))
+			}
+		}
+	}
+	return nil
+}
+
+func (r *Reconciler) validateGuestAccelerators() error {
+
+	a2MachineFamily, n1MachineFamily := r.computeService.MachineTypesList(r.providerSpec.ProjectID, r.providerSpec.Zone, r.Context)
+	machineType := r.providerSpec.MachineType
+	if a2MachineFamily[machineType] != 0 {
+		// a2 family machine - has fixed type and count of GPUs
+		if err := r.checkQuota(a2MachineFamily[machineType]); err != nil {
+			return err
+		} else {
+			return nil
+		}
+	} else if containsString(n1MachineFamily, machineType) {
+		// n1 family machine
+		if err := r.checkQuota(0); err != nil {
+			return err
+		} else {
+			return nil
+		}
+	} else {
+		return nil
```

Comment:
What happens if a user specifies a guest accelerator and they aren't using an A2 or an N1 instance type?

Reply:
googleapi Error gets is returned later when trying [`InstancesInsert()`](https://github.com/openshift/cluster-api-provider-gcp/blob/master/pkg/cloud/gcp/actuators/machine/reconciler.go#L147) looking like this:
`googleapi: Error 400: [<machineType>, <AcceleratorType>] features are not compatible for creating instance., badRequest`

Reply:
I think we can probably return an error here then instead, as we know that this isn't supported. We can write a more informative error for the user I think

Reply:
Something like this ? `InvalidMachineConfiguration(fmt.Sprintf("MachineType %s does not support accelerators.",machinteType))`
Or do we want to be more informative and display accelerator as googleapi error does and maybe specify that only A1 and N1 support accelerators ?

Reply:
I think that's good, I'd add a note on the end as well saying `Only A1 and N1 machine type families support guest acceleartors`